### PR TITLE
[IOTDB-5345] Use the logical clock to identify the snapshot version of IoTConsensus

### PR DESCRIPTION
**Pull request by BUAAserein**
### Please see details in [IOTDB-5345](https://issues.apache.org/jira/browse/IOTDB-5345).

### CODE REVIEW

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 1]
```diff
@@ -379,12 +389,13 @@ private void clearOldSnapshot() {
...
       return;
     }
-    Arrays.sort(versionFiles, Comparator.comparing(File::getName));
-    for (int i = 0; i < versionFiles.length - 1; i++) {
-      try {
-        FileUtils.deleteDirectory(versionFiles[i]);
-      } catch (IOException e) {
-        logger.error("Delete old snapshot dir {} failed", versionFiles[i].getAbsolutePath(), e);
+    for (File file : versionFiles) {
+      if (!file.getName().equals(latestSnapshotId)) {
```
[the remark this answers is not in the document]
Comment (BUAAserein) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java:393]: This method deletes the old files whose names are not the same as the latestSnapshotId. The old method uses string sort which could be wrong such as 2 with 11.

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 2]
```diff
@@ -370,6 +365,21 @@ private String calculateSnapshotPath(String snapshotId, String originalFilePath)
...
+    File[] versionFiles = directory.listFiles((dir, name) -> name.startsWith(SNAPSHOT_DIR_NAME));
+    if (versionFiles == null || versionFiles.length == 0) {
+      return snapShotIndex + 1;
+    }
+    for (File file : versionFiles) {
+      snapShotIndex =
+          Long.max(
+              snapShotIndex, Long.parseLong(file.getName().replaceAll(".*[^\\d](?=(\\d+))", "")));
+    }
+    return snapShotIndex + 1;
```
[the remark this answers is not in the document]
Comment (BUAAserein) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java]: Already changed.

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 3]
```diff
@@ -370,6 +365,21 @@ private String calculateSnapshotPath(String snapshotId, String originalFilePath)
...
+    long snapShotIndex = 0;
+    File directory = new File(storageDir);
+    File[] versionFiles = directory.listFiles((dir, name) -> name.startsWith(SNAPSHOT_DIR_NAME));
+    if (versionFiles == null || versionFiles.length == 0) {
+      return snapShotIndex;
+    }
+    for (File file : versionFiles) {
+      snapShotIndex =
+          Long.max(
+              snapShotIndex, Long.parseLong(file.getName().replaceAll(".*[^\\d](?=(\\d+))", "")));
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java:378]: We may compile this  Pattern once as a class field?

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java:393]: OK

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 4]
```diff
@@ -287,13 +285,10 @@ public DataSet read(IConsensusRequest request) {
 
   public void takeSnapshot() throws ConsensusGroupModifyPeerException {
     try {
-      // TODO: We should use logic clock such as searchIndex rather than wall clock to mark the
-      // snapshot, otherwise there will be bugs in situations where the clock might fall back, such
-      // as CI
+      long latestSnapshotIndex = getLatestSnapshotIndex() + 1;
```
[the remark this answers is not in the document]
Comment (BUAAserein) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java]: Already changed.

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 5]
```diff
@@ -287,13 +285,10 @@ public DataSet read(IConsensusRequest request) {
...
       latestSnapshotId =
```
[the remark this answers is not in the document]
Comment (BUAAserein) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java]: Already changed.

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java (hunk 3 above)
Comment (BUAAserein) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java]: Already changed.

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 6]
```diff
@@ -96,9 +103,10 @@ public void snapshotTest() throws Exception {
             Collections.singletonList(new Peer(dataRegionId, 1, new TEndPoint("0.0.0.0", 9000))));
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:103]: You should call `deletePeer` here to eliminate the relationship between different tests

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 7]
```diff
@@ -96,9 +103,10 @@ public void snapshotTest() throws Exception {
...
 
     Assert.assertTrue(response.isSuccess());
+  }
 
+  public void snapshotTest() throws IOException {
     consensusImpl.triggerSnapshot(dataRegionId);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:109]: You shoud call `createPeer` here

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 8]
```diff
@@ -117,5 +125,48 @@ public void snapshotTest() throws Exception {
     Assert.assertEquals(1, versionFiles2.length);
 
     Assert.assertNotEquals(versionFiles1[0].getName(), versionFiles2[0].getName());
+    for (File file : versionFiles2) {
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:128]: just deletePeer is ok

Comment (BUAAserein) [reply consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java]: Aleady fixed.

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java (hunk 7 above)
Comment (BUAAserein) [reply consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:112]: Aleady fixed.

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java (hunk 6 above)
Comment (BUAAserein) [reply consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:103]: Aleady fixed.

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 9]
```diff
@@ -108,14 +119,55 @@ public void snapshotTest() throws Exception {
...
+    String oldSnapshotId =
+        String.format(
+            "%s_%s_%d",
+            IoTConsensusServerImpl.SNAPSHOT_DIR_NAME,
+            dataRegionId.getId(),
+            System.currentTimeMillis());
+    File snapshotDir = new File(storageDir, "1_1");
+    File oldSnapshotName = new File(snapshotDir, oldSnapshotId);
+    if (oldSnapshotName.exists()) {
+      org.apache.commons.io.FileUtils.deleteDirectory(oldSnapshotName);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:148]: Use `FileUtils.deleteDirectory`

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 10]
```diff
@@ -287,14 +288,11 @@ public DataSet read(IConsensusRequest request) {
 
   public void takeSnapshot() throws ConsensusGroupModifyPeerException {
     try {
-      // TODO: We should use logic clock such as searchIndex rather than wall clock to mark the
-      // snapshot, otherwise there will be bugs in situations where the clock might fall back, such
-      // as CI
-      latestSnapshotId =
+      long newSnapshotIndex = getLatestSnapshotIndex() + 1;
+      newSnapshotFileName =
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java:292]: Sorry for my incorrect review. Maybe it's suitable to be called `newSnapshotDirName`

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 11]
```diff
@@ -108,14 +119,55 @@ public void snapshotTest() throws Exception {
...
 
     Assert.assertNotEquals(versionFiles1[0].getName(), versionFiles2[0].getName());
   }
+
+  public void snapshotUpgradeTest() throws Exception {
+    consensusImpl.createPeer(
+        dataRegionId,
+        Collections.singletonList(new Peer(dataRegionId, 1, new TEndPoint("0.0.0.0", 9000))));
+    consensusImpl.triggerSnapshot(dataRegionId);
+    String oldSnapshotId =
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:139]: oldSnapshotId -> oldSnapshotDirName

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 12]
```diff
@@ -108,14 +119,55 @@ public void snapshotTest() throws Exception {
...
+        String.format(
+            "%s_%s_%d",
+            IoTConsensusServerImpl.SNAPSHOT_DIR_NAME,
+            dataRegionId.getId(),
+            System.currentTimeMillis());
+    File snapshotDir = new File(storageDir, "1_1");
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:145]: snapshotDir -> regionDir

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 13]
```diff
@@ -108,14 +119,55 @@ public void snapshotTest() throws Exception {
...
+    File oldSnapshotName = new File(snapshotDir, oldSnapshotId);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:146]: oldSnapshotName -> oldSnapshotDir

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 14]
```diff
@@ -108,14 +119,55 @@ public void snapshotTest() throws Exception {
...
+    if (oldSnapshotName.exists()) {
+      org.apache.commons.io.FileUtils.deleteDirectory(oldSnapshotName);
+    }
+    if (!oldSnapshotName.mkdirs()) {
+      throw new ConsensusGroupModifyPeerException(
+          String.format("%s: cannot mkdir for snapshot", dataRegionId));
+    }
+    consensusImpl.triggerSnapshot(dataRegionId);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:154]: Maybe you should check  `Assert.assertFalse(oldSnapshotDir.exists())` after this line

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java [hunk 15]
```diff
@@ -108,14 +119,55 @@ public void snapshotTest() throws Exception {
...
+    File dataDir = new File(IoTConsensus.buildPeerDir(storageDir, dataRegionId));
+
+    File[] versionFiles1 =
+        dataDir.listFiles((dir, name) -> name.startsWith(IoTConsensusServerImpl.SNAPSHOT_DIR_NAME));
+    Assert.assertNotNull(versionFiles1);
+    Assert.assertEquals(1, versionFiles1.length);
+
+    consensusImpl.triggerSnapshot(dataRegionId);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/StabilityTest.java:162]: These logic doesn't seem to have anything to do with upgrade

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 16]
```diff
@@ -103,7 +102,8 @@ public class IoTConsensusServerImpl {
   private final IoTConsensusConfig config;
   private final ConsensusReqReader reader;
   private volatile boolean active;
-  private String latestSnapshotId;
+  private String newSnapshotDirName;
+  private final Pattern snapshotIndexPatten = Pattern.compile(".*[^\\d](?=(\\d+))");
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java:106]: static?

Comment (BUAAserein) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java]: Already changed.

**Review by SpriCoder — APPROVED** [review]
LGTM~

**Pull request merged**